### PR TITLE
JBPM-9979 Mark async active nodes and change default active node border color

### PR DESCRIPTION
**Pull request by bxf12315**
…and re-execute it when resuming process flow

**Thank you for submitting this pull request**

**JIRA**: _(please edit the JIRA link if it exists)_ 

[https://issues.redhat.com/browse/JBPM-9945](JBPM-9945)

**referenced Pull Requests**: _(please edit the URLs of referenced pullrequests if they exist)_

* https://github.com/kiegroup/droolsjbpm-integration/pull/2651
* link 2
* link 3 etc.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

**Comment by bxf12315** [comment]
Jenkins run fdb

**Comment by nmirasch** [comment]
@bxf12315 Is this change needed? regarding that ACTIVE_BORDER_COLOR is a configurable property at BC.

**Comment by bxf12315** [comment]
> @bxf12315 Is this change needed? regarding that ACTIVE_BORDER_COLOR is a configurable property at BC.

yes, now default ACTIVE_BORDER_COLOR is red, according to the issue and @lazarotti 's comment, we need to set it to blue, and set error node border to red.

**Comment by nmirasch** [comment]
@bxf12315  I think we need to track this change of the active nodes color in another jira is unrelated to this one "Being able to park process execution on failed script task" and it has to be documented at release notes.

**Comment by nmirasch** [comment]
@bxf12315  looking at this comment, I think it was the expected: https://issues.redhat.com/browse/BAPL-2159?focusedCommentId=19196629&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-19196629. And to me, to be consistent the new ERROR_BORDER_COLOR must be added as a configuration property as the others.

**Comment by bxf12315** [comment]
> @bxf12315 looking at this comment, I think it was the expected: https://issues.redhat.com/browse/BAPL-2159?focusedCommentId=19196629&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-19196629. And to me, to be consistent the new ERROR_BORDER_COLOR must be added as a configuration property as the others.

OK, @lazarotti  means, we need to create a new jira issue to add new configuration property, Could you please add a comment on the issue?

**Comment by bxf12315** [comment]
Jenkins run fdb

**Comment by bxf12315** [comment]
Jenkins run fdb

**Review by nmirasch — COMMENTED** [review]
Code looks good, I just added some minor formatting comments, feel free to review that.
About the commit message maybe something related with the error color configuration addition would reflect better what code does, regarding in the original jira has no reference to this color change introduction.

File: jbpm-wb-common/jbpm-wb-common-api/src/main/java/org/jbpm/workbench/common/preferences/ManagePreferences.java
Comment (jstastny-cz) [inline jbpm-wb-common/jbpm-wb-common-api/src/main/java/org/jbpm/workbench/common/preferences/ManagePreferences.java:70]: please move one line up, closer to the other similar fields initialization.

File: jbpm-wb-common/jbpm-wb-common-api/src/main/java/org/jbpm/workbench/common/preferences/ManagePreferences.java
Comment (jstastny-cz) [inline jbpm-wb-common/jbpm-wb-common-api/src/main/java/org/jbpm/workbench/common/preferences/ManagePreferences.java:114]: I would suggest moving these methods after `setProcessInstanceDiagramActiveNodeBorderColor` to keep similar methods next to each other.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-api/src/main/java/org/jbpm/workbench/pr/service/ProcessRuntimeDataService.java
Comment (jstastny-cz) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-api/src/main/java/org/jbpm/workbench/pr/service/ProcessRuntimeDataService.java:38]: +1

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java
Comment (jstastny-cz) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java:82]:
```suggestion
        when(uiServicesClient.getProcessInstanceImageCustomColor("", null, "", "", "", ""))
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java
Comment (jstastny-cz) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java:106]:
```suggestion
                                                                 "", "")).thenReturn(SVG_WITH_LINK);
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java
Comment (jstastny-cz) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java:112]:
```suggestion
                                                                 "", "");
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java
Comment (jstastny-cz) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java:182]:
```suggestion
                                          completeNodeColor, completeNodeBorderColor, activeNodeBorderColor, "");
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java
Comment (jstastny-cz) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java:186]:
```suggestion
                                                                    activeNodeBorderColor, "");
```

**Review by jstastny-cz — CHANGES_REQUESTED** [review]
There are some formatting issues.

**Comment by bxf12315** [comment]
Jenkins retest this

**Comment by bxf12315** [comment]
Jenkins run fdb

**Comment by bxf12315** [comment]
Jenkins retest this

**Comment by bxf12315** [comment]
Jenkins run fdb

**Comment by barboras7** [comment]
@bxf12315 could you please take a look at the rest of the formatting issues that Jan pointed out? I see two comments that are not marked as "Outdated". Once that is done, I'll check if everything is fixed in the UI and will QE approve the issue if it works as expected.

**Comment by bxf12315** [comment]
> @bxf12315 could you please take a look at the rest of the formatting issues that Jan pointed out? I see two comments that are not marked as "Outdated". Once that is done, I'll check if everything is fixed in the UI and will QE approve the issue if it works as expected.

Done

**Comment by bxf12315** [comment]
Jenkins run fdb

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java
Comment (nmirasch) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java:186]: @bxf12315 makes sense that testProcessInstanceImageCustomColors checks specific values for color configuration. Could you add a variable with  an specific error Node Color and check that value is used i the call. (Like it is done for the other colors)

**Comment by bxf12315** [comment]
Jenkins run fdb

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java
Comment (nmirasch) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java]: @bxf12315  that's it, thanks!

**Comment by nmirasch** [comment]
@bxf12315 There are two missed constants related with the new configuration parameter..

image

the Constants for the new parameter have to be added like the other ones here:
https://github.com/kiegroup/jbpm-wb/blob/75bc89830bb741fed00a09a75d8087d8e74e1c89/jbpm-wb-common/jbpm-wb-common-client/src/main/resources/org/jbpm/workbench/common/client/resources/i18n/Constants.properties#L80

**Comment by bxf12315** [comment]
> @bxf12315 There are two missed constants related with the new configuration parameter..
> 
> image
> 
> the Constants for the new parameter have to be added like the other ones here:
> 
> https://github.com/kiegroup/jbpm-wb/blob/75bc89830bb741fed00a09a75d8087d8e74e1c89/jbpm-wb-common/jbpm-wb-common-client/src/main/resources/org/jbpm/workbench/common/client/resources/i18n/Constants.properties#L80

updated

**Comment by bxf12315** [comment]
Jenkins run fdb

**Comment by bxf12315** [comment]
Jenkins run fdb

**Review by jstastny-cz — COMMENTED** [review]
My previous concerns were resolved. For overall final approval please consider @barboras7 reviews, I won't be available.

**Comment by nmirasch** [comment]
Jenkins run fdb

**Review by nmirasch — APPROVED** [review]

**Comment by barboras7** [comment]
@bxf12315 Could you please provide more information on how to reproduce the issue? I successfully reproduced that the active node color is blue by default. However, I am not able to find a way to cause an error node. Could you please help? I talked with Neus as well, none of us is able to produce an error node.

**Comment by barboras7** [comment]
Hi @bxf12315,
does a custom color for error node work for you? I tried to change both active node color and error node color and only the active node color setting was reflected in the diagram, see attached screenshots. 
Screenshot from 2022-03-07 13-48-27
Screenshot from 2022-03-07 13-05-30
Screenshot from 2022-03-07 13-05-05

**Comment by nmirasch** [comment]
Jenkins run fdb

**Comment by barboras7** [comment]
Hi @bxf12315,
could you please check the functionality for setting custom color of a failed async node? This is still not working for me. I can set a custom color for an active node, but not for failed async node. 

Also, please, take a look at the UI messages provided in the Admin page. We agreed that the issue is not "error node" but "failed async node" and this should be also reflected in what the user sees. 
Screenshot from 2022-03-09 15-03-21

**Comment by bxf12315** [comment]
> Hi @bxf12315, could you please check the functionality for setting custom color of a failed async node? This is still not working for me. I can set a custom color for an active node, but not for failed async node.
> 
> Also, please, take a look at the UI messages provided in the Admin page. We agreed that the issue is not "error node" but "failed async node" and this should be also reflected in what the user sees. Screenshot from 2022-03-09 15-03-21

ok working on it.

**Comment by bxf12315** [comment]
> Hi @bxf12315, could you please check the functionality for setting custom color of a failed async node? This is still not working for me. I can set a custom color for an active node, but not for failed async node.
> 
> Also, please, take a look at the UI messages provided in the Admin page. We agreed that the issue is not "error node" but "failed async node" and this should be also reflected in what the user sees. Screenshot from 2022-03-09 15-03-21

updated, please check it.

**Comment by bxf12315** [comment]
Jenkins run fdb

**Comment by bxf12315** [comment]
Jenkins run fdb

File: jbpm-wb-common/jbpm-wb-common-client/src/main/resources/org/jbpm/workbench/common/client/resources/i18n/Constants.properties
Comment (barboras7) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/resources/org/jbpm/workbench/common/client/resources/i18n/Constants.properties:80]:
```suggestion
ManagePreferences.ProcessInstanceDiagramFailedAsyncNodeBorderColor=Process instance diagram failed async node border color
```

File: jbpm-wb-common/jbpm-wb-common-client/src/main/resources/org/jbpm/workbench/common/client/resources/i18n/Constants.properties
Comment (barboras7) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/resources/org/jbpm/workbench/common/client/resources/i18n/Constants.properties:82]:
```suggestion
ManagePreferences.ProcessInstanceDiagramFailedAsyncNodeBorderColor.Help=Set the color of failed async nodes border at process instance diagram.
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-api/src/main/java/org/jbpm/workbench/pr/service/ProcessImageService.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-api/src/main/java/org/jbpm/workbench/pr/service/ProcessImageService.java:30]:
```suggestion
                                     String failedAsyncNodeBorderColor);
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-api/src/main/java/org/jbpm/workbench/pr/service/ProcessRuntimeDataService.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-api/src/main/java/org/jbpm/workbench/pr/service/ProcessRuntimeDataService.java:39]:
```suggestion
                                                                   String failedAsyncNodeBorderColor);
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/main/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImpl.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/main/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImpl.java:42]:
```suggestion
                                            String failedAsyncNodeBorderColor) {
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/main/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImpl.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/main/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImpl.java:53]:
```suggestion
                                                                                            failedAsyncNodeBorderColor));
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/main/java/org/jbpm/workbench/pr/backend/server/RemoteProcessRuntimeDataServiceImpl.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/main/java/org/jbpm/workbench/pr/backend/server/RemoteProcessRuntimeDataServiceImpl.java:127]:
```suggestion
                                                                          String completedNodeBorderColor, String activeNodeBorderColor, String failedAsyncNodeBorderColor) {
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/main/java/org/jbpm/workbench/pr/backend/server/RemoteProcessRuntimeDataServiceImpl.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/main/java/org/jbpm/workbench/pr/backend/server/RemoteProcessRuntimeDataServiceImpl.java:144]:
```suggestion
                                                                            failedAsyncNodeBorderColor));
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java:184]:
```suggestion
        String failedAsyncNodeBorderColor = "#FF0000";
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java:186]:
```suggestion
                                          completeNodeColor, completeNodeBorderColor, activeNodeBorderColor, failedAsyncNodeBorderColor);
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/test/java/org/jbpm/workbench/pr/backend/server/RemoteProcessImageServiceImplTest.java:190]:
```suggestion
                                                                    activeNodeBorderColor, failedAsyncNodeBorderColor);
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/diagram/ProcessInstanceDiagramPresenter.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/diagram/ProcessInstanceDiagramPresenter.java:181]:
```suggestion
                                            preferences.getProcessInstanceDiagramFailedAsyncNodeBorderColor());
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/instance/diagram/ProcessInstanceDiagramPresenterTest.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/instance/diagram/ProcessInstanceDiagramPresenterTest.java:101]:
```suggestion
    private String failedAsyncNodeBorderColor = "#888885";
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/instance/diagram/ProcessInstanceDiagramPresenterTest.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/instance/diagram/ProcessInstanceDiagramPresenterTest.java:109]:
```suggestion
        when(preferences.getProcessInstanceDiagramFailedAsyncNodeBorderColor()).thenReturn(failedAsyncNodeBorderColor);
```

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/instance/diagram/ProcessInstanceDiagramPresenterTest.java
Comment (barboras7) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/instance/diagram/ProcessInstanceDiagramPresenterTest.java:543]:
```suggestion
        verify(processService).getProcessInstanceDiagramSummary(processInstance.getProcessInstanceKey(), completeNodeColor, completeNodeBorderColor, activeNodeBorderColor, failedAsyncNodeBorderColor);
```

**Review by barboras7 — CHANGES_REQUESTED** [review]

**Comment by bxf12315** [comment]
Jenkins run fdb

**Review by nmirasch — CHANGES_REQUESTED** [review]
@bxf12315 I just left a minor comment about  the renaming of a constant. Overall it  looks good to me and it works fine.

**Comment by nmirasch** [comment]
@bxf12315   could you rename the commit to "JBPM-9979: Mark async active nodes and change default active node border color" to reflect what it is really adding/changing?

File: jbpm-wb-common/jbpm-wb-common-api/src/main/java/org/jbpm/workbench/common/preferences/ManagePreferences.java
Comment (barboras7) [inline jbpm-wb-common/jbpm-wb-common-api/src/main/java/org/jbpm/workbench/common/preferences/ManagePreferences.java]:
```suggestion
        defaultValue.processInstanceDiagramActiveAsyncNodeBorderColor = ACTIVE_ASYNC_BORDER_COLOR;
```

**Review by barboras7 — CHANGES_REQUESTED** [review]

**Comment by nmirasch** [comment]
@bxf12315 the commit message is what needs to be renamed, just to track the feature once it's merged

**Comment by bxf12315** [comment]
> @bxf12315 the commit message is what needs to be renamed, just to track the feature once it's merged

Done

**Comment by nmirasch** [comment]
Jenkins run fdb

**Review by nmirasch — APPROVED** [review]

**Review by barboras7 — APPROVED** [review]

**Pull request merged**
